Poisson_surface_reconstruction_3/include/CGAL/Poisson_mesh_domain_3.h

```diff
@@ -0,0 +1,427 @@
+              .construct_surface_patch_index(Base::create_construct_surface_patch_index(construct_surface_patch_index_)));
+  }
+/// @}
+#ifndef DOXYGEN_RUNNING
+  template<typename CGAL_NP_TEMPLATE_PARAMETERS>
+  static Poisson_mesh_domain_3 create_Poisson_mesh_domain(const CGAL_NP_CLASS& np = parameters::default_values())
+  {
+    static_assert(!parameters::is_default_parameter<CGAL_NP_CLASS, internal_np::function_param_t>::value,
+      "Value for required parameter not found"); static_assert(!parameters::is_default_parameter<CGAL_NP_CLASS,
+internal_np::bounding_object_param_t>::value, "Value for required parameter not found");
+
+    using parameters::get_parameter;
+    return create_Poisson_mesh_domain(parameters::get_parameter(np, internal_np::function_param),
+                                       parameters::get_parameter(np, internal_np::bounding_object_param),
+                                       np);
+  }
+
+  // Overload handling parameters passed with operator=
+  template<typename CGAL_NP_TEMPLATE_PARAMETERS_NO_DEFAULT_1,
+           typename CGAL_NP_TEMPLATE_PARAMETERS_NO_DEFAULT_2,
+           typename ... NP>
+  static Poisson_mesh_domain_3 create_Poisson_mesh_domain(const CGAL_NP_CLASS_1& np1,
+                                                           const CGAL_NP_CLASS_2&  np2,
+                                                           const NP& ... nps)
+  {
+    return create_Poisson_mesh_domain(internal_np::combine_named_parameters(np1, np2, nps...));
+  }
+#endif
+
+  /*
+   * Returns a point in the intersection of the primitive `type`
+   * with some boundary surface.
+   * `Type1` is either `Segment_3`, `Ray_3` or `Line_3`.
+   * The integer `dimension` is set to the dimension of the lowest
+   * dimensional face in the input complex containing the returned point, and
+   * `index` is set to the index to be stored at a mesh vertex lying
+   * on this face.
+   */
+  struct Construct_intersection
+  {
+    Construct_intersection(const Poisson_mesh_domain_3& domain)
+      : r_domain_(domain) {}
+
+    Intersection operator()(const Segment_3& s) const
+    {
+#ifndef CGAL_MESH_3_NO_LONGER_CALLS_DO_INTERSECT_3
+      CGAL_precondition(r_domain_.do_intersect_surface_object()(s) != std::nullopt);
+#endif // NOT CGAL_MESH_3_NO_LONGER_CALLS_DO_INTERSECT_3
+      return this->operator()(s.source(),s.target());
+    }
+
+    Intersection operator()(const Ray_3& r) const
+    {
+      return clip_to_segment(r);
+    }
+
+    Intersection operator()(const Line_3& l) const
+    {
+      return clip_to_segment(l);
+    }
+
+  private:
+    /*
+     * Returns a point in the intersection of `[a,b]` with the surface
+     *  `a` must be the source point, and `b` the out point. It is important
+     * because it drives bisection cuts.
+     * Indeed, the returned point is the first intersection of `[a,b]`
+     * with a subdomain surface.
+     */
+    Intersection operator()(const Point_3& a, const Point_3& b) const
```

Comment:
I'd leave a comment (undocumented) to explain somewhere why we use this class.

Also, couldn't we factorize the implementation with the Labeled mesh domain's (something akin to having an internal function that takes 2 points and a function and does the bisection)? We could probably just have this Construct_intersection inherits the base's too and only redefine the `operator()` (which would call the factorized internal function) and `..._object()`